packages/light-client-extension-helpers/src/background/json-rpc-provider.ts

```diff
@@ -0,0 +1,205 @@
+    const runPromise = Runtime.runPromise(runtime)
+
+    /**
+     * This is a special single-element queue that signals when the daemon is
+     * ready. It is cleared every time the daemon restarts. While a `Deferred`
+     *  would typically be used in this situation, the ability of the daemon
+     * to restart requires a queue that can be cleared and reused.
+     */
+    const daemonReadySignal = yield* Queue.sliding<void>(1)
+
+    /**
+     * A message pubsub for the daemon to publish JSON RPC messages to.
+     * It is the responsibility of the JSON RPC provider to consume these
+     * messages and propagate them with the onMessage callback.
+     */
+    const messagePubSub = yield* PubSub.unbounded<string>()
+
+    /**
+     * An error pubsub for the daemon to publish errors to.
+     * It is the responsibility of the JSON RPC provider to consume these
+     *  errors and propagate them with the onError callback.
+     */
+    const errorPubSub = yield* PubSub.sliding<void>(1)
+
+    /**
+     * A synchronized reference to the current chain. This reference is updated
+     * every time the daemon restarts.
+     *
+     */
+    const chainRef = yield* $(
+      Effect.tryPromise(() => client.addChain(options)),
+      Effect.andThen(SynchronizedRef.make),
+    )
+
+    /**
+     * This daemon process listens for JSON RPC messages, publishing them to
+     * the message pubsub. It also detects errors and publishes those to the
+     * error pubsub.
+     *
+     * If an error occurs with `chain.nextJsonRpcResponse()`, the daemon
+     * restarts. This indicates that the chain is no longer functional.
+     * In response, the daemon reconnects to smoldot using the same
+     * `AddChainParameters` to obtain a new `Chain` object.
+     */
+    const daemon = yield* Effect.gen(function* () {
+      yield* Effect.addFinalizer(() => Queue.takeAll(daemonReadySignal))
+      yield* Effect.addFinalizer(() => PubSub.publish(errorPubSub, undefined))
+
+      const chain = yield* SynchronizedRef.updateAndGetEffect(
+        chainRef,
+        (oldChain) =>
+          $(
+            Effect.tryPromise(() => client.addChain(options)),
+            Effect.tap(() =>
+              $(
+                Effect.try(() => oldChain.remove()),
+                Effect.catchAll(() => Effect.void),
+              ),
+            ),
+          ),
+      )
+
+      // chain is initialized so we can signal that the daemon is ready
+      yield* Queue.offer(daemonReadySignal, undefined)
+
+      yield* $(
+        Effect.tryPromise(() => chain.nextJsonRpcResponse()),
+        Effect.tap(Effect.logDebug),
+        Effect.tap((message) => Queue.offer(messagePubSub, message)),
+      ).pipe(Effect.forever)
+    }).pipe(
+      // by scoping here, we ensure the finalizers are run before the next retry
+      Effect.scoped,
+      Effect.retry(Schedule.spaced("1 second")),
+      Effect.forkDaemon,
+      Effect.interruptible,
+    )
```

Comment:
I still find this effect stuff all very hard to parse; eg just here we have `pipe`, `tryPromise`, `tap`, `Queue.offer`, `scoped`, `forkDaemon` and `interruptible` which are all random effect related things that one needs to read up on to understand what's going on here. And in general, there are a _ton_ of new functions and things just in this file that need to be understood to follow the code, eg `addFInalizer`, `runCallback`, `runPromise`, the whole concept of `Effect.gen` and yielding effect things in a monadic sort of style...

I generally think that we shouldn't use these sorts of libraries so heavily without first getting agreement/buyin from the team, because others need to be willing to learn and udnerstand all of this new stuff to be able to actually review and commit to the code.

I'm not really sure what to suggest, because I don't want to push to rewrite everything. The best "easy" thing that I think could be done to help is to add comments describing exactly what's happening on top of all of these really Effect heavy blocks. If you're keen on Effect, then perhaps it's also worth doing an internal team presentation to highlight some of the reasons it's useful and what this sort of code does?

In general, please do keep in mind going forwards that whatever code you write needs to be readable and understandable by others in the team, else it is unmaintainable.

Reply:
For the record I am trying to experiment with something super basic, but this doesn't work and I have no idea why.

```ts
export const make = async (client: Client, options: AddChainOptions) => {
  let chain = await client.addChain(options)

  const provider = getSyncProvider(async () => {
    try {
      chain.remove()
    } catch (_) {}
    chain = await client.addChain(options)

    return (onMessage, onError) => {
      let connected = true
      ;(async () => {
        while (connected) {
          try {
            const message = await chain.nextJsonRpcResponse()
            onMessage(message)
          } catch (_) {
            onError()
            break
          }
        }
      })()

      const send: JsonRpcConnection["send"] = (message) => {
        try {
          chain.sendJsonRpc(message)
        } catch (_) {}
      }

      const disconnect: JsonRpcConnection["disconnect"] = () => {
        connected = false
        try {
          chain.remove()
        } catch (_) {}
      }

      return {
        send,
        disconnect,
      }
    }
  })

  return provider
}
```

Reply:
Ok we can forget all that, since this just worked. I suppose it was just overengineered.

```ts
export const make = async (client: Client, options: AddChainOptions) => {
  let chain = await client.addChain(options)
  chain.remove()

  const provider = getSyncProvider(async () => {
    chain = await client.addChain(options)

    return (onMessage, onError) => {
      let connected = true
      ;(async () => {
        while (connected) {
          try {
            const message = await chain.nextJsonRpcResponse()
            onMessage(message)
          } catch (_) {
            connected = false
            onError()
            break
          }
        }
      })()

      const send: JsonRpcConnection["send"] = (message) => {
        try {
          chain.sendJsonRpc(message)
        } catch (_) {}
      }

      const disconnect: JsonRpcConnection["disconnect"] = () => {
        connected = false
        try {
          chain.remove()
        } catch (_) {}
      }

      return {
        send,
        disconnect,
      }
    }
  })

  return provider
}
```